[Failing Test]:  :sdks:go:test:ulrValidatesRunner appears to be very flaky at master

### What happened?

Sample failure runs: https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/2996

Also failed on some in-flight PRs.

```
13:19:29 --- PASS: TestEmitParDoAfterGBK (4.50s)
13:19:29 PASS
13:19:29 ok  	github.com/apache/beam/sdks/v2/go/test/regression	124.686s
13:19:29 FAIL
13:19:29 
13:19:29 > Task :sdks:go:test:ulrValidatesRunner FAILED
13:19:29 
13:19:29 FAILURE: Build failed with an exception.
13:19:29 
13:19:29 * Where:
13:19:29 Build file '/home/jenkins/jenkins-slave/workspace/beam_PreCommit_GoPortable_Cron/src/sdks/go/test/build.gradle' line: 152
13:19:29 
13:19:29 * What went wrong:
13:19:29 Execution failed for task ':sdks:go:test:ulrValidatesRunner'.
13:19:29 > Process 'command 'sh'' finished with non-zero exit value 1
13:19:29 
13:19:29 * Try:
13:19:29 > Run with --stacktrace option to get the stack trace.
13:19:29 > Run with --info or --debug option to get more log output.
```

### Issue Failure

Failure: Test is flaky

### Issue Priority

Priority: 1 (unhealthy code / failing or flaky postcommit so we cannot be sure the product is healthy)

### Issue Components

- [ ] Component: Python SDK
- [ ] Component: Java SDK
- [X] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

```
2023/03/31 20:19:00 Prepared job with id: go-testxlang_multi-841-7e5f7f7d-e29d-41a5-85b4-c701695d54d4 and staging token: go-testxlang_multi-841-7e5f7f7d-e29d-41a5-85b4-c701695d54d4
    ptest.go:108: Failed to execute job: 	staging artifacts
        failed to stage /tmp/worker-7-1680293938316973953 in 3 attempts: failed to send chunks for /tmp/worker-7-1680293938316973953
        	caused by:
        chunk send failed
        	caused by:
        EOF; failed to send chunks for /tmp/worker-7-1680293938316973953
        	caused by:
        chunk send failed
        	caused by:
        EOF; failed to send chunks for /tmp/worker-7-1680293938316973953
        	caused by:
        chunk send failed
        	caused by:
        EOF; failed to send chunks for /tmp/worker-7-1680293938316973953
        	caused by:
        chunk send failed
        	caused by:
        EOF
--- FAIL: TestXLang_Multi (12.99s)
```

Same thing for some of the others. But it's very odd that it's happening to the Python runner and not Flink/Spark/Samza.

Feels sort of like a grpc thing, but again, not sure why it's only started recently.

--- a/sdks/go/pkg/beam/artifact/stage.go
+++ b/sdks/go/pkg/beam/artifact/stage.go
@@ -159,11 +159,11 @@ func Stage(ctx context.Context, client jobpb.LegacyArtifactStagingServiceClient,
 	}
 	stagedHash, err := stageChunks(stream, fd)
 	if err != nil {
-		stream.CloseAndRecv() // ignore error
-		return nil, errors.Wrapf(err, "failed to send chunks for %v", filename)
+		_, errClose := stream.CloseAndRecv()
+		return nil, errors.Wrapf(err, "failed to send chunks for %v; close error: %v", filename, errClose)
 	}
-	if _, err := stream.CloseAndRecv(); err != nil && err != io.EOF {
-		return nil, errors.Wrapf(err, "failed to close stream for %v", filename)
+	if resp, err := stream.CloseAndRecv(); err != nil && err != io.EOF {
+		return nil, errors.Wrapf(err, "failed to close stream for %v; response: %v", filename, resp)
 	}
 	if hash != stagedHash {
 		return nil, errors.Errorf("unexpected SHA256 for sent chunks for %v: %v, want %v", filename, stagedHash, hash)
@@ -188,7 +188,11 @@ func stageChunks(stream jobpb.LegacyArtifactStagingService_PutArtifactClient, r 
 					},
 				},
 			}
-			if err := stream.Send(chunk); err != nil {
+			err := stream.Send(chunk)
+			if err == io.EOF {
+				return "", err
+			}
+			if err != nil {
 				return "", errors.Wrap(err, "chunk send failed")
 			}
 		}

--- a/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
+++ b/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
@@ -62,14 +62,15 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 	defer func() {
 		if retErr != nil {
 			log.Error(ctx, "StageViaPortableApi error: ", retErr)
-		}
-		if err := stream.CloseSend(); err != nil {
-			log.Error(ctx, "StageViaPortableApi CloseSend error: ", err)
-			retErr = err
+			if err := stream.CloseSend(); err != nil {
+				log.Error(ctx, "StageViaPortableApi CloseSend error: ", err)
+				retErr = err
+			}
 		}
 	}()
 
 	if err := stream.Send(&jobpb.ArtifactResponseWrapper{StagingToken: st}); err != nil {
+		log.Error(ctx, "StageViaPortableApi staging token send error: ", err)
 		return err
 	}
 
@@ -79,6 +80,7 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 			return nil
 		}
 		if err != nil {
+			log.Error(ctx, "StageViaPortableApi Recv error: ", err)
 			return err
 		}
 
@@ -123,7 +125,7 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtifactRetrievalServiceClient) error {
 	fd, err := os.Open(filename)
 	if err != nil {
-		return err
+		return errors.Wrapf(err, "unable to open file %v", filename)
 	}
 	defer fd.Close()
 
@@ -152,6 +154,10 @@ func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtif
 				Response: &jobpb.ArtifactResponseWrapper_GetArtifactResponse{
 					GetArtifactResponse: &jobpb.GetArtifactResponse{},
 				}})
+
+			if sendErr != nil {
+				log.Infof(context.TODO(), "error on Final Send for %v: %v", filename, sendErr)
+			}
 			return sendErr
 		}
 

--- a/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
+++ b/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
@@ -19,6 +19,7 @@ import (
 	"context"
 	"io"
 	"os"
+	"strings"
 	"time"
 
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/artifact"
@@ -42,8 +43,7 @@ func Stage(ctx context.Context, id, endpoint, binary, st string) (retrievalToken
 	}
 	defer cc.Close()
 
-	err = StageViaPortableApi(ctx, cc, binary, st)
-	if err == nil {
+	if err := StageViaPortableApi(ctx, cc, binary, st); err == nil {
 		return "", nil
 	} else {
 		log.Warnf(ctx, "unable to stage with PortableAPI: %v; falling back to legacy", err)
@@ -53,25 +53,36 @@ func Stage(ctx context.Context, id, endpoint, binary, st string) (retrievalToken
 }
 
 func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st string) (retErr error) {
-	client := jobpb.NewArtifactStagingServiceClient(cc)
+	const attempts = 3
+	var failures []string
+	for {
+		err := stageFiles(ctx, cc, binary, st)
+		if err == nil {
+			return nil // success!
+		}
+		if err != nil {
+			failures = append(failures, err.Error())
+		}
+		if len(failures) > attempts {
+			return errors.Errorf("failed to stage artifacts for token %v in %v attempts: %v", st, attempts, strings.Join(failures, ";\n"))
+		}
+	}
+}
 
-	stream, err := client.ReverseArtifactRetrievalService(context.Background())
+func stageFiles(ctx context.Context, cc *grpc.ClientConn, binary, st string) error {
+	client := jobpb.NewArtifactStagingServiceClient(cc)
+	stream, err := client.ReverseArtifactRetrievalService(ctx)
 	if err != nil {
 		return err
 	}
 	defer func() {
-		if retErr != nil {
-			log.Error(ctx, "StageViaPortableApi error: ", retErr)
-			if err := stream.CloseSend(); err != nil {
-				log.Error(ctx, "StageViaPortableApi CloseSend error: ", err)
-				retErr = err
-			}
+		if err := stream.CloseSend(); err != nil {
+			log.Error(ctx, "StageViaPortableApi CloseSend error: ", err)
 		}
 	}()
 
 	if err := stream.Send(&jobpb.ArtifactResponseWrapper{StagingToken: st}); err != nil {
-		log.Error(ctx, "StageViaPortableApi staging token send error: ", err)
-		return err
+		return errors.Wrapf(err, "failed to send staging token")
 	}
 
 	for {
@@ -80,7 +91,6 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 			return nil
 		}
 		if err != nil {
-			log.Error(ctx, "StageViaPortableApi Recv error: ", err)
 			return err
 		}
 
@@ -101,21 +111,19 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 			// TODO(https://github.com/apache/beam/issues/21459): Legacy Type URN. If requested, provide the binary.
 			// To be removed later in 2022, once thoroughly obsolete.
 			case graphx.URNArtifactGoWorker:
-				if err := StageFile(binary, stream); err != nil {
+				if err := stageFile(binary, stream); err != nil {
 					if err == io.EOF {
-						log.Infof(ctx, "failed to Go worker binary %v due to server side close.", binary)
 						continue // so we can get the real error from stream.Recv.
 					}
-					return errors.Wrap(err, "failed to stage Go worker binary")
+					return errors.Wrapf(err, "failed to stage Go worker binary: %v", binary)
 				}
 			case graphx.URNArtifactFileType:
 				typePl := pipepb.ArtifactFilePayload{}
 				if err := proto.Unmarshal(request.GetArtifact.Artifact.TypePayload, &typePl); err != nil {
 					return errors.Wrap(err, "failed to parse artifact file payload")
 				}
-				if err := StageFile(typePl.GetPath(), stream); err != nil {
+				if err := stageFile(typePl.GetPath(), stream); err != nil {
 					if err == io.EOF {
-						log.Infof(ctx, "failed to stage file %v due to server side close.", typePl.GetPath())
 						continue // so we can get the real error from stream.Recv.
 					}
 					return errors.Wrapf(err, "failed to stage file %v", typePl.GetPath())
@@ -131,7 +139,7 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 	}
 }
 
-func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtifactRetrievalServiceClient) error {
+func stageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtifactRetrievalServiceClient) error {
 	fd, err := os.Open(filename)
 	if err != nil {
 		return errors.Wrapf(err, "unable to open file %v", filename)
@@ -149,7 +157,6 @@ func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtif
 					},
 				}})
 			if sendErr == io.EOF {
-				log.Infof(context.TODO(), "StageFile error on Send of len %v for %v: %v", n, filename, sendErr)
 				return sendErr
 			}
 
@@ -164,10 +171,6 @@ func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtif
 				Response: &jobpb.ArtifactResponseWrapper_GetArtifactResponse{
 					GetArtifactResponse: &jobpb.GetArtifactResponse{},
 				}})
-
-			if sendErr != nil {
-				log.Infof(context.TODO(), "StageFile error on Final Send for %v: %v", filename, sendErr)
-			}
 			return sendErr
 		}
 

--- a/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
+++ b/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
@@ -103,7 +103,7 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 			case graphx.URNArtifactGoWorker:
 				if err := StageFile(binary, stream); err != nil {
 					if err == io.EOF {
-						log.Infof(ctx, "failed to Go worker binary %v:", binary, err)
+						log.Infof(ctx, "failed to Go worker binary %v due to server side close.", binary)
 						continue // so we can get the real error from stream.Recv.
 					}
 					return errors.Wrap(err, "failed to stage Go worker binary")
@@ -115,7 +115,7 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 				}
 				if err := StageFile(typePl.GetPath(), stream); err != nil {
 					if err == io.EOF {
-						log.Infof(ctx, "failed to stage file %v:", typePl.GetPath(), err)
+						log.Infof(ctx, "failed to stage file %v due to server side close.", typePl.GetPath())
 						continue // so we can get the real error from stream.Recv.
 					}
 					return errors.Wrapf(err, "failed to stage file %v", typePl.GetPath())

--- a/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
+++ b/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
@@ -24,6 +24,7 @@ import (
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/artifact"
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/runtime/graphx"
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/internal/errors"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/log"
 	jobpb "github.com/apache/beam/sdks/v2/go/pkg/beam/model/jobmanagement_v1"
 	pipepb "github.com/apache/beam/sdks/v2/go/pkg/beam/model/pipeline_v1"
 	"github.com/apache/beam/sdks/v2/go/pkg/beam/util/grpcx"
@@ -41,8 +42,11 @@ func Stage(ctx context.Context, id, endpoint, binary, st string) (retrievalToken
 	}
 	defer cc.Close()
 
-	if err := StageViaPortableApi(ctx, cc, binary, st); err == nil {
+	err = StageViaPortableApi(ctx, cc, binary, st)
+	if err == nil {
 		return "", nil
+	} else {
+		log.Warnf(ctx, "unable to stage with PortableAPI: %v; falling back to legacy", err)
 	}
 
 	return StageViaLegacyApi(ctx, cc, binary, st)

--- a/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
+++ b/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
@@ -102,6 +102,10 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 			// To be removed later in 2022, once thoroughly obsolete.
 			case graphx.URNArtifactGoWorker:
 				if err := StageFile(binary, stream); err != nil {
+					if err == io.EOF {
+						log.Infof(ctx, "failed to Go worker binary %v:", binary, err)
+						continue // so we can get the real error from stream.Recv.
+					}
 					return errors.Wrap(err, "failed to stage Go worker binary")
 				}
 			case graphx.URNArtifactFileType:
@@ -110,7 +114,12 @@ func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st st
 					return errors.Wrap(err, "failed to parse artifact file payload")
 				}
 				if err := StageFile(typePl.GetPath(), stream); err != nil {
+					if err == io.EOF {
+						log.Infof(ctx, "failed to stage file %v:", typePl.GetPath(), err)
+						continue // so we can get the real error from stream.Recv.
+					}
 					return errors.Wrapf(err, "failed to stage file %v", typePl.GetPath())
+
 				}
 			default:
 				return errors.Errorf("request has unexpected artifact type %s", typeUrn)

--- a/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
+++ b/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
@@ -140,11 +140,12 @@ func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtif
 					},
 				}})
 			if sendErr == io.EOF {
+				log.Infof(context.TODO(), "StageFile error on Send of len %v for %v: %v", n, filename, sendErr)
 				return sendErr
 			}
 
 			if sendErr != nil {
-				return errors.Wrap(sendErr, "chunk send failed")
+				return errors.Wrap(sendErr, "StageFile chunk send failed")
 			}
 		}
 
@@ -156,7 +157,7 @@ func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtif
 				}})
 
 			if sendErr != nil {
-				log.Infof(context.TODO(), "error on Final Send for %v: %v", filename, sendErr)
+				log.Infof(context.TODO(), "StageFile error on Final Send for %v: %v", filename, sendErr)
 			}
 			return sendErr
 		}

--- a/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
+++ b/sdks/go/pkg/beam/runners/universal/runnerlib/stage.go
@@ -52,13 +52,22 @@ func Stage(ctx context.Context, id, endpoint, binary, st string) (retrievalToken
 	return StageViaLegacyApi(ctx, cc, binary, st)
 }
 
-func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st string) error {
+func StageViaPortableApi(ctx context.Context, cc *grpc.ClientConn, binary, st string) (retErr error) {
 	client := jobpb.NewArtifactStagingServiceClient(cc)
 
 	stream, err := client.ReverseArtifactRetrievalService(context.Background())
 	if err != nil {
 		return err
 	}
+	defer func() {
+		if retErr != nil {
+			log.Error(ctx, "StageViaPortableApi error: ", retErr)
+		}
+		if err := stream.CloseSend(); err != nil {
+			log.Error(ctx, "StageViaPortableApi CloseSend error: ", err)
+			retErr = err
+		}
+	}()
 
 	if err := stream.Send(&jobpb.ArtifactResponseWrapper{StagingToken: st}); err != nil {
 		return err
@@ -128,6 +137,9 @@ func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtif
 						Data: data[:n],
 					},
 				}})
+			if sendErr == io.EOF {
+				return sendErr
+			}
 
 			if sendErr != nil {
 				return errors.Wrap(sendErr, "chunk send failed")
@@ -152,7 +164,7 @@ func StageFile(filename string, stream jobpb.ArtifactStagingService_ReverseArtif
 func StageViaLegacyApi(ctx context.Context, cc *grpc.ClientConn, binary, st string) (retrievalToken string, err error) {
 	client := jobpb.NewLegacyArtifactStagingServiceClient(cc)
 
-	files := []artifact.KeyedFile{artifact.KeyedFile{Key: "worker", Filename: binary}}
+	files := []artifact.KeyedFile{{Key: "worker", Filename: binary}}
 
 	md, err := artifact.MultiStage(ctx, client, 10, files, st)
 	if err != nil {

--- a/sdks/go/pkg/beam/runners/universal/runnerlib/stage_test.go
+++ b/sdks/go/pkg/beam/runners/universal/runnerlib/stage_test.go
@@ -1,0 +1,173 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package runnerlib
+
+import (
+	"context"
+	"io"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/runtime/graphx"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/util/protox"
+	jobpb "github.com/apache/beam/sdks/v2/go/pkg/beam/model/jobmanagement_v1"
+	pipepb "github.com/apache/beam/sdks/v2/go/pkg/beam/model/pipeline_v1"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/test/bufconn"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/testing/protocmp"
+)
+
+func initDummyServer(ctx context.Context, t *testing.T) (*grpc.ClientConn, *dummyArtifactServer) {
+	t.Helper()
+	const buffsize = 1024 * 1024
+	l := bufconn.Listen(buffsize)
+
+	server := grpc.NewServer()
+	das := &dummyArtifactServer{
+		t: t,
+	}
+	jobpb.RegisterArtifactStagingServiceServer(server, das)
+	jobpb.RegisterLegacyArtifactStagingServiceServer(server, das)
+
+	go func() {
+		server.Serve(l)
+	}()
+
+	t.Cleanup(func() {
+		server.Stop()
+		l.Close()
+	})
+
+	clientConn, err := grpc.DialContext(ctx, "", grpc.WithContextDialer(func(ctx context.Context, _ string) (net.Conn, error) {
+		return l.DialContext(ctx)
+	}), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+	if err != nil {
+		t.Fatal("couldn't create bufconn grpc connection:", err)
+	}
+	return clientConn, das
+}
+
+func TestPortableArtifactStaging(t *testing.T) {
+	ctx, cancelFn := context.WithTimeout(context.Background(), 5*time.Second)
+	t.Cleanup(cancelFn)
+
+	cc, das := initDummyServer(ctx, t)
+
+	das.wantToken = "token"
+	das.reqFile = "stage.go"
+
+	err := StageViaPortableApi(ctx, cc, "reqFile", das.wantToken)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+type dummyArtifactServer struct {
+	jobpb.UnimplementedArtifactStagingServiceServer
+	jobpb.UnimplementedLegacyArtifactStagingServiceServer
+
+	t *testing.T
+
+	wantToken, reqFile string
+}
+
+func (das *dummyArtifactServer) ReverseArtifactRetrievalService(stream jobpb.ArtifactStagingService_ReverseArtifactRetrievalServiceServer) error {
+	msg, err := stream.Recv()
+	if err == io.EOF {
+		das.t.Error("ReverseArtifactRetrievalService.Recv: unexpected EOF")
+		return nil
+	}
+	if err != nil {
+		das.t.Errorf("ReverseArtifactRetrievalService.Recv: unexpected error %v", err)
+		return err
+	}
+	if got, want := msg.GetStagingToken(), das.wantToken; got != want {
+		das.t.Errorf("ReverseArtifactRetrievalService.Recv: unexpected token %v, want %v", got, want)
+	}
+
+	// Check artifact Resolve Requests.
+	wantArts := []*pipepb.ArtifactInformation{
+		{
+			TypeUrn:     "dummy",
+			TypePayload: []byte("dummy"),
+			RoleUrn:     "dummy",
+			RolePayload: []byte("dummy"),
+		},
+	}
+
+	if err := stream.Send(&jobpb.ArtifactRequestWrapper{
+		Request: &jobpb.ArtifactRequestWrapper_ResolveArtifact{
+			ResolveArtifact: &jobpb.ResolveArtifactsRequest{
+				Artifacts: wantArts,
+			},
+		},
+	}); err != nil {
+		das.t.Fatalf("unexpected err on artifact resolve request Send: %v", err)
+	}
+
+	msg, err = stream.Recv()
+	if err != nil {
+		das.t.Fatalf("unexpected err on artifact resolve response Recv: %v", err)
+	}
+	resolve := msg.GetResolveArtifactResponse()
+	if resolve == nil {
+		das.t.Fatalf("unexpected err on artifact resolve response Resp, got %T", msg.GetResponse())
+	}
+	gotArts := resolve.GetReplacements()
+	if d := cmp.Diff(wantArts, gotArts, protocmp.Transform()); d != "" {
+		das.t.Errorf("diff on artifact resolve response (-want, +got):\n%v", d)
+	}
+
+	typePl := protox.MustEncode(&pipepb.ArtifactFilePayload{
+		Path: das.reqFile,
+	})
+
+	// Check file upload requests
+	if err := stream.Send(&jobpb.ArtifactRequestWrapper{
+		Request: &jobpb.ArtifactRequestWrapper_GetArtifact{
+			GetArtifact: &jobpb.GetArtifactRequest{
+				Artifact: &pipepb.ArtifactInformation{
+					TypeUrn:     graphx.URNArtifactFileType,
+					TypePayload: typePl,
+				},
+			},
+		},
+	}); err != nil {
+		das.t.Fatalf("unexpected err on get artifact request Send: %v", err)
+	}
+
+	msg, err = stream.Recv()
+	if err != nil {
+		das.t.Fatalf("unexpected err on get artifact response Recv: %v", err)
+	}
+
+	if msg.GetIsLast() {
+		das.t.Fatalf("unexpected IsLast on get artifact response Recv: %v", prototext.Format(msg))
+	}
+	msg, err = stream.Recv()
+	if err != nil {
+		das.t.Fatalf("unexpected err on artifact resolve response Recv: %v", err)
+	}
+	if !msg.GetIsLast() {
+		das.t.Fatalf("want IsLast on get artifact response Recv: %v", prototext.Format(msg))
+	}
+
+	return nil
+}

